File: src/main/java/com/vjiki/music/mapper/ShortMapper.java
package com.vjiki.music.mapper;

import com.vjiki.music.dto.ShortResponse;
import com.vjiki.music.entity.Song;

public final class ShortMapper {

    private ShortMapper() {
    }

    public static ShortResponse toShortResponse(Song song) {
        return toShortResponse(song, false, false, song.getLikesCount(), song.getDislikesCount());
    }

    public static ShortResponse toShortResponse(Song song, boolean isLiked, boolean isDisliked) {
        return toShortResponse(song, isLiked, isDisliked, song.getLikesCount(), song.getDislikesCount());
    }

    public static ShortResponse toShortResponse(Song song,
                                                boolean isLiked,
                                                boolean isDisliked,
                                                Long likesCount,
                                                Long dislikesCount) {
        return ShortResponse.builder()
                .id(song.getId() == null ? null : song.getId().toString())
                .artist(SongMapper.extractArtist(song.getArtists()))
                .audioUrl(SongMapper.extractUrl(song.getAudioUrls()))
                .cover(SongMapper.extractUrl(song.getCoverUrls()))
                .title(song.getTitle())
                .videoUrl(SongMapper.extractUrl(song.getVideoUrls()))
                .type(song.getType())
                .isLiked(isLiked)
                .isDisliked(isDisliked)
                .likesCount(likesCount == null ? 0L : likesCount)
                .dislikesCount(dislikesCount == null ? 0L : dislikesCount)
                .build();
    }
}

File: bin/main/com/vjiki/music/mapper/ShortMapper.kt
package com.vjiki.music.mapper

import com.vjiki.music.dto.ShortResponse
import com.vjiki.music.entity.Song
import org.springframework.stereotype.Component

@Component
object ShortMapper {
    fun Song.toShortResponse(
        isLiked: Boolean = false,
        isDisliked: Boolean = false,
        likesCount: Long = this.likesCount,
        dislikesCount: Long = this.dislikesCount
    ): ShortResponse {
        return ShortResponse(
            id = id.toString(),
            artist = extractArtist(artists),
            audioUrl = extractUrl(audioUrls),
            cover = extractUrl(coverUrls),
            title = title,
            videoUrl = extractUrl(videoUrls),
            type = type,
            isLiked = isLiked,
            isDisliked = isDisliked,
            likesCount = likesCount,
            dislikesCount = dislikesCount
        )
    }

    private fun extractArtist(artists: Map<String, List<String>>?): String? {
        return artists?.get("default")?.firstOrNull()
    }

    private fun extractUrl(urls: Map<String, String>?): String? {
        return urls?.get("default")
    }
}

